### PR TITLE
schedulers: add role config to shuffleRegionScheduler

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
Fix https://github.com/pingcap/pd/issues/2177

### What is changed and how it works?
Add role configuration to shuffle-region-scheduler.

Use it to balance learners:

**1. add shuffle-region-scheduler**
```
pd-ctl scheduler add shuffle-region-scheduler
```
or
```
curl -XPOST -d '{"name": "shuffle-region-scheduler"}' http://pd:2379/pd/api/v1/schedulers
```

**2. config roles**
```
pd-ctl scheduler config shuffle-region-scheduler set-roles learner
```
or
```
curl -XPOST -d '["learner"]'  http://pd:2379/pd/api/v1/scheduler-config/shuffle-region-scheduler/roles
```

### Check List 

Tests 
 - Unit test
 - Manual test

Code changes
 - Has HTTP API interfaces change (Don't forget to [update API document](https://github.com/pingcap/pd/blob/master/docs/development.md#updating-api-documentation))

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### CODE REVIEW

File: server/schedulers/shuffle_region_config.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+func (conf *shuffleRegionSchedulerConfig) IsRoleAllow(role string) bool {
+	conf.RLock()
+	defer conf.RUnlock()
+	return slice.AnyOf(conf.Roles, func(i int) bool { return conf.Roles[i] == role })
+}
+
+func (conf *shuffleRegionSchedulerConfig) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	router := mux.NewRouter()
+	router.HandleFunc("/roles", conf.handleGetRoles).Methods("GET")
```
Comment (shafreeck) [inline server/schedulers/shuffle_region_config.go]: Is this the root path of  PD's API? Maybe add the PD API prefix is better?

Comment (disksing) [reply server/schedulers/shuffle_region_config.go]: no, it is a sub router to be registered under `scheduler_config/shuffle-region-scheduler/`.

File: server/schedulers/shuffle_region.go [hunk 2]
```diff
@@ -125,31 +125,48 @@ func (s *shuffleRegionScheduler) Schedule(cluster opt.Cluster) []*operator.Opera
...
 }
 
-func (s *shuffleRegionScheduler) scheduleAddPeer(cluster opt.Cluster, filter filter.Filter) *metapb.Peer {
-	stores := cluster.GetStores()
+func (s *shuffleRegionScheduler) scheduleAddPeer(cluster opt.Cluster, region *core.RegionInfo, oldPeer *metapb.Peer) *metapb.Peer {
+	var scoreGuard filter.Filter
+	if cluster.IsPlacementRulesEnabled() {
+		scoreGuard = filter.NewRuleFitFilter(s.GetName(), cluster, region, oldPeer.GetStoreId())
+	} else {
+		scoreGuard = filter.NewDistinctScoreFilter(s.GetName(), cluster.GetLocationLabels(), cluster.GetRegionStores(region), cluster.GetStore(oldPeer.GetStoreId()))
```
Comment (rleungx) [inline server/schedulers/shuffle_region.go]: Is it different from the original behavior?

Comment (disksing) [reply server/schedulers/shuffle_region.go]: yes, different. Priviously it does not consider any constraints.

File: tools/pd-ctl/pdctl/command/scheduler.go [hunk 3]
```diff
@@ -531,3 +550,40 @@ func deleteStoreFromSchedulerConfig(cmd *cobra.Command, schedulerName string, ar
...
+		return
+	}
+	var roles []string
+	fields := strings.Split(strings.ToLower(args[0]), ",")
+	for _, f := range fields {
+		if f != "" {
+			roles = append(roles, f)
+		}
+	}
+	b, _ := json.Marshal(roles)
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/scheduler.go]: Do we need to handle this error?

Comment (disksing) [reply tools/pd-ctl/pdctl/command/scheduler.go]: it should not return any error as it only contains basic types.

**Comment by disksing** [comment]
CI failure cased by https://github.com/pingcap/pd/issues/2231

**Pull request merged**